### ISSUE TITLE
AudioDXE - dump directly to Platforms.xml

### ISSUE THREAD

**Issue by roddy20** [acidanthera/bugtracker#1947]
it is not a bug, but a good idea for the future. We can save audiocodec dump to xml matching with Platforms.xml format. Then load it to AppleALC by some kind of resource convertor. Result - any (or almost any) HDA codec will work without using of precompiled large and buggy database

**Comment by mikebeaton** [comment]
AudioDxe includes code to handle input channels, but doesn't currently configure them - in fact the opposite, we fairly agressively try to configure output on any channel that could be one. So you'd at least need some new input channel set up code, and some optional user config as to which discovered channels were i or o.

Then I believe there's some additional setup, which AudioDxe he doesn't need to do, to set up speakers and headphones to switch depending on what is plugged in.

Finally, we'd need to set up the codec (and/or pass the setup) specifically so that sound can be driven by Apple HDA, not just by AudioDxe. Just for myself, I'm not currently aware of the requirements of Apple HDA, and exactly what that translation would look like, and how feasible it would be.

**Comment by Core-i99** [comment]
But audiodxe can make an audio codec dump. Isn't it possible to convert that to platforms.xml?
There's already codec graph which can convert Linux codec dumps to a graph. (https://github.com/Core-i99/Codec-Graph)
I used codec graph to make platforms.xml for my layout id, and that was pretty useful. Maybe when I have some time I'll give it a try to start from the existing coed graph code and let it make platforms.xml

**Comment by roddy20** [comment]
Yes, it is possible and we do it this way and we did it a lot of times
I talk about automatic EFI-utility to save the dump in XML without Linux and 3rd party tools

**Comment by Core-i99** [comment]
> Yes, sure it is possible and we do it this way and we did it a lot of times
> I talk about some kind of EFI-utility to save the dump in XML without Linux and 3rd party tools and import it directly to AppleALC at system startup (automatic or semi-automatic). Why do we use linux dump? We may save it directly as XML

I don't think windows support dumping the codec that advanced as Linux or audiodxe can do. Injecting a codec dump isn't a good idea, applealc will be needed to recompiled and applelac will get way bigger then it needs to be. 

The way I currently want to do it :
Dump codec with audiodxe > comvert it to platforms.xml, layout.xml and pinconfigs > recompile applealc

**Comment by mikebeaton** [comment]
> Yes, sure it is possible and we do it this way and we did it a lot of times I talk about some kind of EFI-utility to save the dump in XML without Linux and 3rd party tools and import it directly to AppleALC at system startup (automatic or semi-automatic). Why do we use linux dump? We may save it directly as XML

Isn't the codec dump done by OpenCore SysReport already almost exactly the same as the Linux codec dump?

**Comment by mikebeaton** [comment]
> > Yes, sure it is possible and we do it this way and we did it a lot of times
> > I talk about some kind of EFI-utility to save the dump in XML without Linux and 3rd party tools and import it directly to AppleALC at system startup (automatic or semi-automatic). Why do we use linux dump? We may save it directly as XML
> 
> I don't think windows support dumping the codec that advanced as Linux or audiodxe can do. Injecting a codec dump isn't a good idea, applealc will be needed to recompiled and applelac will get way bigger then it needs to be.
> 
> The way I currently want to do it : Dump codec with audiodxe > generate platforms.xml, layout.xml and pinconfigs > recompile applealc

Isn't the codec dump done by OpenCore SysReport already almost exactly the same as the Linux codec dump?

**Comment by mikebeaton** [comment]
@roddy20 AudioDxe already parses all the codec widgets and works out for itself how to connect them up. You're suggesting leveraging that existing setup and passing to AppleALC, yes?

**Comment by roddy20** [comment]
> Isn't the codec dump done by OpenCore SysReport already almost exactly the same as the Linux codec dump?
Yes, it is almost the same

**Comment by roddy20** [comment]
> AudioDxe already parses all the codec widgets and works out for itself how to connect them up. You're suggesting leveraging that existing setup and passing to AppleALC, yes?

I suggest to save the dump to XML in addition to plain text, as the first step, it will be possible to use it without converting by 3rd party tools, and maybe it wiill be possible to do it automatically in the future

**Comment by mikebeaton** [comment]
@roddy20 Yourself and @Core-i99 seem to be suggesting that it is a well-known and pretty much automatic process to convert from a Linux-format codec dump to the full layout and configs needed by AppleALC to get AppleHDA to drive any card correctly. I think a lot of this knowledge is lost on old forum posts! I have worked on AudioDxe, and other (mainly front end) bits of OpenCore, but I have not worked on AppleALC (till adding some compatibility improvements to `alc-verb`, just recently) and I am not aware of how this automatic conversion process would work. All the links I've seen, followed and read seem to imply that this conversion is a very involved process, with a lot of manual steps. Please could you clarify?

**Comment by roddy20** [comment]
[GetDumpXML 1.0 3.zip](https://github.com/acidanthera/bugtracker/files/8030285/GetDumpXML.1.0.3.zip)
it is not my code
program receives dump from VoodooHDA and builds Platform.XML, why it is not possible for AudioDXE?

**Comment by roddy20** [comment]
we need to get PIN Speaker for example, check node connection list, find mixer node, check node connection list, find DAC, write the sequence to XML. if mixer has volume control, add it too. I can do it manually but I am not so good in programming

**Comment by roddy20** [comment]
> @roddy20  Please could you clarify?

I think the desired XML structure can be seen from the sources
I dont know exactly how to add this to AudioDXE, or maybe it will be a new tool, but it does not seem to be impossible

**Comment by roddy20** [comment]
[DumpXML_Analog Devices AD1988B.zip](https://github.com/acidanthera/bugtracker/files/8030371/DumpXML_Analog.Devices.AD1988B.zip)
the result for my codec
it is not perfect, but it really works

**Comment by roddy20** [comment]
[Codec1.txt.zip](https://github.com/acidanthera/bugtracker/files/8030510/Codec1.txt.zip)
dump from Opencore system report for the same codec

**Comment by roddy20** [comment]
Node 0x12 [Pin Complex] wcaps 0x40018D: Stereo Amp-Out
  Amp-Out caps: ofs=0x00, nsteps=0x00, stepsize=0x00, mute=1
  Amp-Out vals: [0x80 0x80]
  Pincap 0x0000373F: IN OUT HP Trigger Detect
  Pin Default 0x01014010: [Jack] Line Out at Ext Rear
    Conn = 1/8, Color = Green
    DefAssociation = 0x1, Sequence = 0x0
Pin-ctls: 0x02:
  Connection: 1
     0x29

Node 0x29 [Audio Mixer] wcaps 0x200103: Stereo Amp-In
  Amp-In caps: ofs=0x00, nsteps=0x00, stepsize=0x00, mute=1
  Amp-In vals: [0x80 0x80] [0x80 0x80]
  Connection: 2
     0x04 0x21

Node 0x04 [Audio Output] wcaps 0x405: Stereo Amp-Out
  Amp-Out caps: ofs=0x27, nsteps=0x27, stepsize=0x05, mute=0
  Amp-Out vals: [0x27 0x27]
  Converter: stream=0, channel=0

**Comment by roddy20** [comment]
0x12 0x29 0x04 is the first path from Out to DAC 18 41 4 in XML etc
nsteps=0x27, stepsize=0x05, mute=0 is for Volume Slider , Apple calls it PublishVolume
nsteps=0x00, stepsize=0x00, mute=1 is for Mute Button, MuteInputAMP

**Comment by roddy20** [comment]
for the first attempt we may skip 2.1/5.1/7.1 cоnfigurations and try one speaker, one headphones as outputs, one mic as input, it is enough to see if it works

**Comment by Core-i99** [comment]
> [GetDumpXML 1.0 3.zip](https://github.com/acidanthera/bugtracker/files/8030285/GetDumpXML.1.0.3.zip)
> it is not my code
> program receives dump from VoodooHDA and builds Platform.XML, why it is not possible for AudioDXE?

Is that project on github? Or where are the releases of it?

**Comment by roddy20** [comment]
> Is that project on github? Or where are the releases of it?

it was created here https://applelife.ru/threads/getdumpxml-applehda-zvuk-avtomaticheskoe-postroenie-fajla-platforms-xml.2942828/ for and it was forgotten with new versions of AppleALC and VoodooHDA. And injection of custom Platforms.xml to /S/L/E/AppleHDA.kext became impossible since Mojave or Catalina
